src/utils/daos.c

```diff
@@ -970,13 +994,79 @@ fs_op_hdlr(struct cmd_args_s *ap)
 		if (ap->src == NULL || ap->dst == NULL) {
 			fprintf(stderr, "a source and destination path "
 				"must be provided\n");
+			D_GOTO(out, rc = RC_PRINT_HELP);
 		} else {
 			rc = fs_copy_hdlr(ap);
 		}
 		break;
+	case FS_SET_ATTR:
+	case FS_GET_ATTR:
+	case FS_RESET_ATTR:
+	case FS_RESET_CHUNK_SIZE:
+	case FS_RESET_OCLASS:
+		if (ap->path != NULL) {
+			struct duns_attr_t      dattr = {0};
+
+			if (ap->dfs_path) {
+				fprintf(stderr, "can't specify file or dir in "
+					"path and dfs_path at the same time\n");
+				return EINVAL;
+			}
+
+			rc = duns_resolve_path(ap->path, &dattr);
+			/** we could be creating a new file, so try dirname */
+			if (rc == ENOENT && op == FS_SET_ATTR) {
+				parse_filename_dfs(ap->path, &name,
+						   &dir_name);
+
+				rc = duns_resolve_path(dir_name, &dattr);
+			}
+			if (rc) {
+				fprintf(stderr, "could not resolve pool &"
+					" container by path: %d %s %s\n",
+					rc, strerror(rc), ap->path);
+				D_GOTO(out, rc);
+			}
+
+			ap->type = dattr.da_type;
+			uuid_copy(ap->p_uuid, dattr.da_puuid);
+			uuid_copy(ap->c_uuid, dattr.da_cuuid);
+
+			if (name) {
+				if (dattr.da_rel_path) {
+					D_ASPRINTF(ap->dfs_path, "%s/%s",
+						   dattr.da_rel_path, name);
+					free(dattr.da_rel_path);
+				} else {
+					D_ASPRINTF(ap->dfs_path, "/%s", name);
+				}
+			} else {
+				if (dattr.da_rel_path) {
+					D_STRNDUP(ap->dfs_path,
+						  dattr.da_rel_path, PATH_MAX);
+					free(dattr.da_rel_path);
+				} else {
+					D_STRNDUP(ap->dfs_path, "/", 1);
+				}
+			}
+			if (ap->dfs_path == NULL)
+				D_GOTO(out, rc = ENOMEM);
+		} else {
+			ARGS_VERIFY_PUUID(ap, out, rc = RC_PRINT_HELP);
+			ARGS_VERIFY_CUUID(ap, out, rc = RC_PRINT_HELP);
+		}
+		rc = fs_dfs_hdlr(ap);
+		if (rc)
+			D_GOTO(out, rc);
+		break;
 	default:
 		break;
 	}
+
+out:
+	D_FREE(dir_name);
+	D_FREE(name);
+	D_FREE(ap->dfs_path);
```

Comment:
If we put this `D_FREE(ap->dfs_path);` in the args_free function, then it will also be free'ed on any other miscellaneous error condition, including errors before fs_op_hdlr is called.